audbcards/core/dataset.py

```diff
@@ -274,6 +292,283 @@ def version_link(self) -> str:
+                else:
+                    # e.g. {'f': 'female', 'm': 'male'}
+                    mappings = '✓'
+
+        data.append(mappings)
+        data_dict['Mappings'] = mappings
+
+        # data = [x if x != [] else "" for x in data]
+        # data = [x for x in data if x is not None]
+        # data = ', '.join(['"'+x+'"' for x in data])
+        return data_dict
+
+    @property
+    def scheme_info(self) -> dict:
+
+        db = self.header
+        scheme_info = {}
+
+        if len(db.schemes) > 0:
+            has_minimums = any(
+                [db.schemes[s].minimum is not None for s in db.schemes]
+            )
+            has_maximums = any(
+                [db.schemes[s].maximum is not None for s in db.schemes]
+            )
+            has_labels = any(
+                [db.schemes[s].labels is not None for s in db.schemes]
+            )
+            has_mappings = any(
+                [
+                    isinstance(db.schemes[s].labels, (str, dict))
+                    for s in db.schemes
+                ]
+            )
+
+        columns = ['ID', 'Dtype']
+        header_line = '   :header: ID,Dtype'
+        if has_minimums:
+            header_line += ',Min'
+            columns.append('Min')
+        if has_maximums:
+            header_line += ',Max'
+            columns.append('Max')
+        if has_labels:
+            header_line += ',Labels'
+            columns.append('Labels')
+        if has_mappings:
+            header_line += ',Mappings'
+            columns.append('Mappings')
+
+        scheme_info['header_line'] = header_line
+        scheme_info['has_minimums'] = has_minimums
+        scheme_info['has_maximums'] = has_maximums
+        scheme_info['has_labels'] = has_labels
+        scheme_info['has_mappings'] = has_mappings
+
+        scheme_info['columns'] = columns
+        return scheme_info
+
+    @property
+    def dataset_schemes(self) -> list:
+
+        db = self.header
+        dataset_schemes = []
+        for scheme_id in db.schemes:
+            dataset_scheme = self._scheme_to_list(scheme_id)
+            dataset_schemes.append(dataset_scheme)
+
+        cols = self.scheme_info['columns']
+        data = pd.DataFrame.from_dict(dataset_schemes)[cols]
+        filter = data.applymap(lambda d: d == [])
+        data.mask(filter, other='', inplace=True)
+        scheme_data = np.array(data).tolist()
+        return scheme_data
+
+
+class Datacard(object):
+
+    def __init__(self, dataset: Dataset):
+
+        self._dataset = dataset
+
+    @property
+    def props(self) -> dict:
+        props = self._dataset.properties()
+        return props
+
+    def _render_template(self):
+
+        t_dir = os.path.join(os.path.dirname(__file__), 'templates')
+        environment = jinja2.Environment(loader=jinja2.FileSystemLoader(t_dir),
+                                         trim_blocks=True)
+        environment.filters.update(zip=zip,
```

Comment:
Where comes `zip` from, I don't see it defined inside this function?

Reply:
This is just the builtin for parallel traversals like in e.g. `zip(list1,list2)`. This is required as a jinja filter but not present by default.  See it  used in `datacard_tables.j2` where it is used to iterate over `tables, types and columns`  in parallel.

The second filter trim whitespace `tw`  is used to delete empty table rows, see `datacard_schemes.j2`.

[This](https://stackoverflow.com/questions/5208252/ziplist1-list2-in-jinja2) post from sov does a similar thing but uses the global namespace.

Reply:
Ah, yes, of cause, the build-in `zip`. I did not imagine that something like this needs to be handed to another function, but I now understand it. I will make another suggestion how to add a comment here and resolve this conversation.